unittests/unit/test_ConstraintSolver.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2011-2019, The DART development contributors
+ * All rights reserved.
+ *
+ * The list of contributors can be found at:
+ *   https://github.com/dartsim/dart/blob/master/LICENSE
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <iostream>
+
+#include <gtest/gtest.h>
+
+#include "dart/constraint/ConstraintBase.hpp"
+#include "dart/constraint/ConstraintSolver.hpp"
+#include "dart/simulation/World.hpp"
+
+#include "TestHelpers.hpp"
+
+using namespace dart;
+using namespace dart::neural;
+
+TEST(ConstraintSolver, SIMPLE)
+{
+  std::shared_ptr<simulation::World> world
+      = std::make_shared<simulation::World>();
```

Comment:
Besides just compiling, we also want to check that there aren't any crashes due to a null-pointer anywhere (which the compiler can't fully check for you at compile time), so it's worth setting up a test that actually executes the code you're changing.

You may want to actually create some interesting constraints in this world object.

For example, you could have two objects that are colliding, which creates at least one CollisionConstraint? We can then build on that test as we add features for differentiation through these decomposed functions.

I'll add comments down below with the specific things we should check.